### ISSUE TITLE
Having an @-sign in a let-binding removes the rest of the name, but not in type members

### ISSUE THREAD

**Issue by abelbraaksma** [dotnet/fsharp#7784]
An `@` sign in a let-binding removes the rest of the name, and replaces it with a number (typically: the line number).

**Repro steps**
Consider this (1):

```f#
type Tests() =
    [<Test>]
    static member``This is a test with @sign in it`` () = Assert.True(true)
```

And this (2):

```f#
module Tests=
    [<Test>]
    let ``This is a test with @sign in it`` () = Assert.True(true)
```

(1) compiles to `Tests.``This is a test with @sign in it`` `
(2) compiles to `Tests.``This is a test with @35`` ` (depending on line number)

**Expected behavior**

Both let bindings and members should be treated the same way. After all, they can both be public identifiers and consumed from a library.

A compiler warning (as per F# spec), but other than that, the name should be left untouched. I'm aware that F# compiles lambdas etc to to methods with special names including the `@`-sign, but the risk of name clashes should be for the programmer, right?

**Actual behavior**

The name is garbled and shortened, which looks surprising and, especially with the feature of the doubly backticked long names, it is a pity this cannot be left untouched.

**Known workarounds**

Don't use public let-bindings for tests (or whenever you need to use a `@`-sign in the name).

**Related information**

I would assume this is due to the internals of F# that typically only apply to internal F# generated identifiers and for some reason it thinks this name is also a generated name (which it clearly isn't, so I'm just wondering why it gets the lambda-treatment).

Seen on recent F# versions, may also happen on older versions, I didn't check.

**Comment by abelbraaksma** [comment]
As an additional workaround, it turns out one can (ab)use `CompiledNameAttribute` for this purpose. However, this showed another bug: this attribute behaves rather odd when applied to a name that contains an `@`-sign. By removing the at-sign from the original name, and adding it instead to the `CompiledName` attribute solved this. 

Still, I don't think module level names should be truncated to begin with.

**Comment by tboby** [comment]
I traced this back to 
https://github.com/dotnet/fsharp/blob/4330fa1b60d8c7fd9c1dcb7dd8155fef5f025c61/src/Compiler/Checking/CheckPatterns.fs#L193

Where it relies on the presence of `@` to detect if something is compiler generated, even if `isCompGen` is false in `vFlags`.

**Comment by abelbraaksma** [comment]
@tboby, yes, I figured (at the time) that this has something to do with the reason that `@`-signs are used for compiler-generated internals. It seems very shaky that we both allow that `@`-sign in names, and at the same time have a test that says `if there's an @-sign its internal`.

Feels very shaky. @dsyme, we use the attribute `CompilerGeneratedAttribute`, isn't that way better to use to deduct whether a name is compiler generated? If that breaks code, alternatively, we can disallow `@`-signs completely (or raise a warning).

**Comment by vzarytovskii** [comment]
Tagging @0101 who were tinkering with compiler generated attributes recently.
